### DISCUSSION TITLE
How to Configure LazyVim to Use an LSP Server Inside a Docker Container?

### DISCUSSION THREAD

**Discussion by ravihammond** [LazyVim/LazyVim#4889] [General]
I'm new to Neovim, Lazy.nvim, and LazyVim (though I have a few years of experience with vanilla Vim), and I also have no prior experience with LSPs. So, please explain this like I’m a complete beginner!

Here’s my setup:
- The code I’m working on is dependent on a Docker container. All the necessary dependencies are installed via the Dockerfile, and I run the code using this Docker image.
- However, I edit the code locally using my local LazyVim installation and configuration.

The problem I’m facing is that when I open my Python files in LazyVim, the LSP flags numerous errors because it can’t find the dependencies. This makes sense since the dependencies are installed inside the Docker container, not on my local machine.

**My question:**

How can I set this up so that my local LazyVim uses an LSP server running inside the Docker container, while still using my local Neovim as the LSP client?

**Comment by kylejbrk** [comment]
@ravihammond did you ever find a solution to this?

**Reply by ChillyWall** [reply]
I have a solution, you can see whether it fits your needs.

**Comment by ChillyWall** [comment]
@ravihammond It's easy to let LazyVim use a lsp server inside docker container, you just need to change the cmd of that lsp's config.

Add a file in your own config's plugin directory, for example, pyright.lua (I assume you are using pyright), with content below. Replace `/path/to/node`, `container-name`, and `/path/to/pyright-langserver` with the real values on your machine. 

> Note: Make sure the current user can execute docker commands without `sudo`

```lua
-- ~/.config/nvim/lua/plugins/pyright.lua
return {
  "neovim/nvim-lspconfig",
  opts = {
    servers = {
      pyright = {
        cmd = {
          "docker",
          "exec",
          "-i",
          "container-name",
          "bash",
          "-c",
          "source",
          "/root/.bashrc",
          "&&", 
          "pyright-langserver",
          "--stdio",
        },
      },
    },
  },
}
```

This is just an example, you can modify as you like. (If you use this, make sure you have node and pyright-langserver in you PATH.)

Then you have to make pyright-langserver be able to read your code.

You can create a volume and put your project into it. You can just use soft links to make sure the path is the same between your host and container. (It's difficult to transform the path, for pyright doesn't support this by its own.)

For example, run your container with `-v /home/username/workspace:/workspace`, then run `ln -s /workspace /home/username/workspace` in container.
